restapi/src/main/java/io/stargate/web/docsapi/dao/Paginator.java

```diff
@@ -0,0 +1,106 @@
+package io.stargate.web.docsapi.dao;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.stargate.db.datastore.Row;
+import io.stargate.web.docsapi.exception.DocumentAPIRequestException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+
+/**
+ * This class is in charge of keeping the data of the Docs API pagination process. It's not only a
+ * stateful component, but a state machine with a initial state. It's created on the REST layer and
+ * passed through to the DocumentService that is in charge of its changes. After being populated its
+ * `documentPageState` can be passed by the REST layer to the client as a JSON string.
+ */
+public class Paginator {
+
+  private final ObjectMapper mapper = new ObjectMapper();
+
+  public final int dbPageSize;
+  public final int docPageSize;
+
+  private boolean initialState = true;
+  private ByteBuffer currentDbPageState; // keeps track of the DB page state while querying the DB
+  private DocumentSearchPageState documentPageState = null;
+
+  public Paginator(String pageStateParam, int pageSizeParam, int dbPageSize) throws IOException {
+    docPageSize = Math.max(1, pageSizeParam);
+    if (docPageSize > DocumentDB.MAX_PAGE_SIZE) {
+      throw new DocumentAPIRequestException("The parameter `page-size` is limited to 20.");
+    }
+
+    this.dbPageSize = dbPageSize;
+
+    if (pageStateParam != null) {
+      byte[] decodedBytes = Base64.getDecoder().decode(pageStateParam);
+      documentPageState = mapper.readValue(decodedBytes, DocumentSearchPageState.class);
+    }
+
+    this.currentDbPageState = (documentPageState != null) ? documentPageState.getPageState() : null;
+  }
+
+  public String getDocumentPageStateAsString() throws JsonProcessingException {
```

Comment:
is it possible that this method is called N times? If yes, then maybe we should save the result of (pre-compute):
```
 if (documentPageState != null) {
      byte[] pagingJson = mapper.writeValueAsBytes(documentPageState);
      return Base64.getEncoder().encodeToString(pagingJson);
    }
```
in the field and just return it here?

Reply:
Technically it'll only be called once per HTTP request as of now